src/v/storage/index_state.h

```diff
@@ -139,14 +139,14 @@ struct index_state
     // the index.
     std::optional<model::timestamp> broker_timestamp{std::nullopt};
 
-    // The offset of the first compactible record in the segment. This may not
-    // necessarily correspond to one of the values in the indexing vectors.
+    // The number of compactible records appended to the segment. This may not
+    // necessarily indicate the exact number of compactible records, e.g. if
+    // the segment was truncated, the count will remain the same. As such, this
+    // value may be an overestimate of the exact number of compactible records.
     //
-    // Special values:
-    // - nullopt: indicates no compactible batches are in the segment.
-    // - disabled: indicates this index was written in a version that didn't
-    //             support this field, and we can't conclude anything.
-    tristate<model::offset> first_compactible_offset{std::nullopt};
+    // Returns std::nullopt if this index was written in a version that didn't
+    // support this field, and we can't conclude anything.
+    std::optional<size_t> num_compactible_records_appended{0};
```

Comment:
is the zero value default here meaningless in that we either decode some value or its set to nullopt when decoding from older version? we never encode nullopt, right? so i guess i'm wondering if the default should be nullopt?

Reply:
The zero here is the equivalent of `std::nullopt` in the previous version: in either case, this is the default value for a fresh `index_state` created in this version.

In the same way we were using tristate {std::nullopt} before, the default value is the indicator that, at any given point after creation, we've indexed some records and still haven't seen any compactible records, which tells us that there are actually no records in this segment. Versus the now std::nullopt case, which means we just don't know because the index was written in an older version.

Reply:
ahh right of course. thanks